src/pybind/mgr/orchestrator/_interface.py

```diff
@@ -973,6 +973,40 @@ def to_json(self) -> dict:
             del out[e]
         return out
 
+    def to_dict(self) -> dict:
+        out: Dict[str, Any] = OrderedDict()
+        out['daemon_type'] = self.daemon_type
+        out['daemon_id'] = self.daemon_id
+        out['hostname'] = self.hostname
+        out['container_id'] = self.container_id
+        out['container_image_id'] = self.container_image_id
+        out['container_image_name'] = self.container_image_name
+        out['container_image_digests'] = self.container_image_digests
+        out['memory_usage'] = self.memory_usage
+        out['memory_request'] = self.memory_request
+        out['memory_limit'] = self.memory_limit
+        out['version'] = self.version
+        out['status'] = self.status.value if self.status is not None else None
+        out['status_desc'] = self.status_desc
+        if self.daemon_type == 'osd':
+            out['osdspec_affinity'] = self.osdspec_affinity
+        out['is_active'] = self.is_active
+        out['ports'] = self.ports
+        out['ip'] = self.ip
+
+        for k in ['last_refresh', 'created', 'started', 'last_deployed',
+                  'last_configured']:
+            if getattr(self, k):
+                out[k] = datetime_to_str(getattr(self, k))
+
+        if self.events:
+            out['events'] = [e.to_dict() for e in self.events]
+
+        empty = [k for k, v in out.items() if v is None]
+        for e in empty:
+            del out[e]
+        return out
```

Comment:
And same for this, sorry!
```py
return {
  'daemon_type': ...
```

If we drop support for Python 3.6 (I think CentOS 8 supports 3.8), we could use [`dataclasses.asdict()`](https://docs.python.org/3.7/library/dataclasses.html#dataclasses.asdict).

Reply:
This refactoring actually requires a lot of code change in the cephadm code. There are a number of tests that are impacted by this. It would be better if we take this as an improvement in a separate PR as those changes would be unrelated to the purpose that this PR is serving.

Reply:
You mean the dataclasses? Because a refactoring like the one suggested shouldn't impact any existing tests given this is a new method and I don't see any unit test testing this.

BTW, no need to use an OrderedDict here as [dict's in Py3.6 are by default ordered](https://bugs.python.org/issue27350):
```suggestion
    def to_dict(self) -> dict:
        return {
          'daemon_type': self.daemon_type,
          'daemon_id': self.daemon_id,
          'hostname': self.hostname,
          ...
        }
```

Reply:
There are tests in cephadm and orchestrator which are affected : https://github.com/ceph/ceph/blob/4029981e459e0f2474a5b3c7fe802e444a871596/src/pybind/mgr/cephadm/tests/test_cephadm.py#L103

https://github.com/ceph/ceph/blob/4029981e459e0f2474a5b3c7fe802e444a871596/src/pybind/mgr/orchestrator/tests/test_orchestrator.py#L84

https://github.com/ceph/ceph/blob/4029981e459e0f2474a5b3c7fe802e444a871596/src/pybind/mgr/cephadm/tests/test_spec.py#L127

Since in the approach that you are suggesting, we are not removing the attributes that are 'None', so we need to change the expected schemas in all these tests. And in-fact in each case there are 5-10 values that are expected as None.

Reply:
Ok, I now see where it fails. My bad. Sorry @aaSharma14. I don't think it's a good practice to have a dynamic schema 'because some values are None' (in that case, the Schema should define that a property can either have a value or be None).

In any case, let's go with it. We have enough code to clean-up in dashboard without having to deal with orchestrator too.